src/lib/transition_frontier/transition_frontier.ml

```diff
@@ -1 +1,274 @@
+   *     b) apply the ledger builder diff to the new mask
+   *     c) form the breadcrumb and node records
+   *   2) add the node to the table
+   *   3) add the successor_hashes entry to the parent node
+   *   4) move the root if the path to the new node is longer than the max length
+   *     a) calculate the distance from the new node to the parent
+   *     b) if the distance is greater than the max length:
+   *       I  ) find the immediate successor of the old root in the path to the
+   *            longest node and make it the new root
+   *       II ) find all successors of the other immediate successors of the old root
+   *       III) remove the old root and all of the nodes found in (II) from the table
+   *       IV ) merge the old root's merkle mask into the root ledger
+   *   5) set the new node as the best tip if the new node has a greater length than
+   *      the current best tip
+   *)
+  let add_exn t transition =
+    let protocol_state = External_transition.protocol_state transition in
+    let hash = Protocol_state.hash protocol_state in
+    let root_node = Hashtbl.find_exn t.table t.root in
+    let best_tip_node = Hashtbl.find_exn t.table t.best_tip in
+    let parent_hash = Protocol_state.previous_state_hash protocol_state in
+    let parent_node = Hashtbl.find_exn t.table parent_hash in
+    (* 1.a *)
+    let mask = Merkle_mask.derive (Breadcrumb.mask parent_node.breadcrumb) in
+    (* 1.b *)
+    Merkle_mask.apply mask (External_transition.ledger_builder_diff transition) ;
+    (* 1.c *)
+    let node =
+      { breadcrumb= {Breadcrumb.transition; mask}
+      ; successor_hashes= []
+      ; length= parent_node.length + 1 }
+    in
+    (* 2 *)
+    assert (`Ok = Hashtbl.add t.table ~key:hash ~data:node) ;
+    (* 3 *)
+    Hashtbl.set t.table ~key:parent_hash
+      ~data:
+        { parent_node with
+          successor_hashes= hash :: parent_node.successor_hashes } ;
+    (* 4.a *)
+    let distance_to_parent = root_node.length - node.length in
+    (* 4.b *)
+    if distance_to_parent > max_length then (
+      (* 4.b.I *)
+      let new_root_hash = List.hd_exn (path t node.breadcrumb) in
+      (* 4.b.II *)
+      let garbage_immediate_successors =
+        List.filter root_node.successor_hashes ~f:(fun succ_hash ->
+            not (State_hash.equal succ_hash new_root_hash) )
+      in
+      (* 4.b.III *)
+      let garbage =
+        t.root
+        :: List.concat
+             (List.map garbage_immediate_successors ~f:(successor_hashes_rec t))
+      in
+      t.root <- new_root_hash ;
+      List.iter garbage ~f:(Hashtbl.remove t.table) ;
+      (* 4.b.IV *)
+      Merkle_ledger.merge t.root_ledger (Breadcrumb.mask root_node.breadcrumb) ) ;
+    (* 5 *)
+    let best_tip_node = Hashtbl.find_exn t.table t.best_tip in
+    if node.length > best_tip_node.length then t.best_tip <- hash ;
+    node.breadcrumb
+end
+
+(*
+let%test "transitions can be added and interface will work at any point" =
```

Comment:
Can you put this in a `let%test_module` and just do `let disabled_test () = (* ... *)` so it can compile when we dune runtest?

Reply:
Sure.